zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs

```diff
@@ -0,0 +1,134 @@
+                acc * (1.expr() - value.expr() * nonempty_witness.clone().expr())
+            }),
+        );
+
+        Self {
+            is_zero,
+            nonempty_witness,
+        }
+    }
+
+    pub(crate) fn expr(&self) -> Expression<F> {
+        self.is_zero.expr()
+    }
+
+    pub(crate) fn assign(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        values: [F; N],
+    ) -> Result<F, Error> {
+        let is_zero =
+            if let Some(inverse) = values.iter().find_map(|value| Option::from(value.invert())) {
+                self.nonempty_witness
+                    .assign(region, offset, Value::known(inverse))?;
+                F::zero()
+            } else {
+                F::one()
+            };
+        self.is_zero.assign(region, offset, Value::known(is_zero))?;
+
+        Ok(is_zero)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::test_util::*;
+    use super::*;
+    use crate::evm_circuit::util;
+    use eth_types::*;
+    use halo2_proofs::halo2curves::bn256::Fr;
+    use halo2_proofs::plonk::Error;
+
+    #[test]
+    fn test_batched_iszero() {
+        const N: usize = 32;
+        #[derive(Clone)]
+        /// all(n.cells) == 0
+        struct IsZeroGadgetTestContainer<F> {
+            z_gadget: BatchedIsZeroGadget<F, N>,
+            n: util::Word<F>,
+        }
+
+        impl<F: Field> MathGadgetContainer<F> for IsZeroGadgetTestContainer<F> {
+            const NAME: &'static str = "BatchedIsZeroGadget";
+
+            fn configure_gadget_container(cb: &mut ConstraintBuilder<F>) -> Self {
+                let n = cb.query_word();
+                let z_gadget = BatchedIsZeroGadget::<F, N>::construct(
+                    cb,
+                    n.cells
+                        .iter()
+                        .map(|cell| cell.expr())
+                        .collect::<Vec<Expression<F>>>()
+                        .try_into()
+                        .unwrap(),
+                );
+                cb.require_equal("Input must be all 0", z_gadget.expr(), 1.expr());
+                IsZeroGadgetTestContainer { z_gadget, n }
+            }
+
+            fn assign_gadget_container(
+                &self,
+                input_words: &[Word],
+                region: &mut CachedRegion<'_, '_, F>,
+            ) -> Result<(), Error> {
+                let n = input_words[0];
+                let offset = 0;
+
+                self.n.assign(region, offset, Some(n.to_le_bytes()))?;
+                self.z_gadget.assign(
+                    region,
+                    0,
+                    n.to_le_bytes().map(|byte| F::from(byte as u64)),
+                )?;
+
+                Ok(())
+            }
+        }
+
+        test_math_gadget_container::<Fr, IsZeroGadgetTestContainer<Fr>>(vec![Word::from(0)], true);
+
+        test_math_gadget_container::<Fr, IsZeroGadgetTestContainer<Fr>>(vec![Word::from(1)], false);
```

Comment:
Could use a test case where only a single value wasn't `0` or something like that.

Reply:
1 is that case, but it a little bit special as the bit pos is in the first pos. will add a number like 1024 for same purpose test.

Reply:
Oh I see, I didn't notice that the byte values of the word were passed into the gadget. Is there a reason to do it like that? It's a bit confusing and doesn't seem to make testing easier (harder to specify different values).

Reply:
That's because the WORD (U256) is by design to be a unified input param for all gadgets, as it can be easily changed to any other value type. The Pro is we have an all in one test framework, however, the Cons is a little bit confusing for some cases which do not take WORD. It well fits that old all-gadgets-in-one-file math_gadget.rs. Now I think it still fine, but yes, it can be managed better as each file has own code space.